subsys/bluetooth/audio/pacs.c

```diff
@@ -627,92 +730,249 @@ static int pac_notify_snk(void)
+		.snk = sys_cpu_to_le16(snk_available_contexts),
+		.src = sys_cpu_to_le16(src_available_contexts),
+	};
+	int err;
 
+	err = pacs_gatt_notify(conn, BT_UUID_PACS_AVAILABLE_CONTEXT, pacs_svc.attrs,
+				  &context, sizeof(context));
+	if (err != 0 && err != -ENOTCONN) {
+		LOG_WRN("Available Audio Contexts notify failed: %d", err);
 		return err;
 	}
 
-	get_pac_records(&src_pacs_list, &read_buf);
+	return 0;
+}
 
-	err = bt_gatt_notify_uuid(NULL, BT_UUID_PACS_SRC, pacs_svc.attrs,
-				  read_buf.data, read_buf.len);
+static int supported_contexts_notify(struct bt_conn *conn)
+{
+	struct bt_pacs_context context = {
+		.snk = sys_cpu_to_le16(snk_supported_contexts),
+		.src = sys_cpu_to_le16(src_supported_contexts),
+	};
+	int err;
+
+	if (!IS_ENABLED(CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE)) {
+		return -ENOTSUP;
+	}
+
+	err = pacs_gatt_notify(conn, BT_UUID_PACS_SUPPORTED_CONTEXT, pacs_svc.attrs,
+				  &context, sizeof(context));
 	if (err != 0 && err != -ENOTCONN) {
-		LOG_WRN("PACS notify failed: %d", err);
+		LOG_WRN("Supported Audio Contexts notify failed: %d", err);
+
+		return err;
 	}
+	return 0;
+}
 
-	k_sem_give(&read_buf_sem);
+void pacs_gatt_notify_complete_cb(struct bt_conn *conn, void *user_data)
+{
+	/* Notification done, clear bit and reschedule work */
+	atomic_clear(&notify_rdy);
+	k_work_submit(&deferred_nfy_work);
+}
+
+static int pacs_gatt_notify(struct bt_conn *conn,
+			    const struct bt_uuid *uuid,
+			    const struct bt_gatt_attr *attr,
+			    const void *data,
+			    uint16_t len)
+{
+	int err;
+	struct bt_gatt_notify_params params;
+
+	memset(&params, 0, sizeof(params));
+	params.uuid = uuid;
+	params.attr = attr;
+	params.data = data;
+	params.len  = len;
+	params.func = pacs_gatt_notify_complete_cb;
+
+	/* Mark notification in progress */
+	atomic_set(&notify_rdy, 1);
+
+	err = bt_gatt_notify_cb(conn, &params);
+	if (err != 0) {
+		atomic_clear(&notify_rdy);
+	}
 
 	if (err == -ENOTCONN) {
 		return 0;
 	} else {
 		return 0;
 	}
 }
-#endif /* CONFIG_BT_PAC_SRC */
 
-static int pacs_changed(enum bt_audio_dir dir)
+static void notify_cb(struct bt_conn *conn, void *data)
 {
+	struct pacs_client *client = &clients[bt_conn_index(conn)];
+	struct bt_conn_info info;
 	int err = 0;
 
-	if (dir == BT_AUDIO_DIR_SINK) {
-#if defined(CONFIG_BT_PAC_SNK)
-		err = pac_notify_snk();
-#endif /* CONFIG_BT_PAC_SNK */
-	} else {
-#if defined(CONFIG_BT_PAC_SRC)
-		err = pac_notify_src();
-#endif /* CONFIG_BT_PAC_SRC */
+	LOG_DBG("Notify cb");
+
+	err = bt_conn_get_info(conn, &info);
+	if (err != 0) {
+		LOG_ERR("Failed to get conn info: %d", err);
+		return;
 	}
 
-	return err;
+	if (info.state != BT_CONN_STATE_CONNECTED) {
+		/* Not connected */
+		return;
+	}
+
+	/* Check if we have unverified notifications in progress */
+	if (atomic_get(&notify_rdy)) {
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_PAC_SNK_NOTIFIABLE) &&
+	    atomic_test_bit(client->flags, FLAG_SINK_PAC_CHANGED)) {
+		printk("Notifying Sink PAC\n");
```

Comment:
Avoid using `printk`, use logging subsystem (LOG_DBG) instead. Applies in multiple places

Reply:
Yes, this was indeed some nasty leftover :)